libs/parallelism/algorithms/include/hpx/parallel/algorithms/starts_ends_with.hpp

```diff
@@ -0,0 +1,132 @@
+//  Copyright (c) 2020 ETH Zurich
+//  Copyright (c) 2014 Grant Mercer
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+/// \file parallel/algorithms/starts_ends_with.hpp
+
+#pragma once
+
+#include <hpx/config.hpp>
+#include <hpx/execution/algorithms/detail/predicates.hpp>
+#include <hpx/executors/execution_policy.hpp>
+#include <hpx/iterator_support/traits/is_iterator.hpp>
+#include <hpx/parallel/algorithms/detail/dispatch.hpp>
+#include <hpx/parallel/algorithms/detail/distance.hpp>
+#include <hpx/parallel/container_algorithms/equal.hpp>
+#include <hpx/parallel/container_algorithms/mismatch.hpp>
+#include <hpx/parallel/util/detail/algorithm_result.hpp>
+#include <hpx/parallel/util/invoke_projected.hpp>
+#include <hpx/parallel/util/projection_identity.hpp>
+
+#include <algorithm>
+#include <cstddef>
+#include <iterator>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace hpx { namespace parallel { inline namespace v1 {
```

Comment:
Should the base implementation be here or should they be in their respective `container_algorithms/starts_with.hpp` and `container_algorithms/ends_with.hpp`?

Reply:
Let's have them in separate headers to go along with the conventions applied for the other algorithms.

Reply:
> Let's have them in separate headers to go along with the conventions applied for the other algorithms.

Ok, I misunderstood what you were asking. Let's have the base implementations here. Could those be made independent of the ranges headers, though? I would like to avoid for the algorithms headers to depend on ranges headers.

Reply:
Oh ok sure, can I use the parallel headers of mismatch and equal?

Unless as mentioned in the other comment, I'll add our own implementation which dosen't use mismatch and equal.

Reply:
> Oh ok sure, can I use the parallel headers of mismatch and equal?

Sure. I don't see any problems with this. We do this for other algorithms as well (e.g. copy uses for_each, etc.)